# Pridictive Analysis of Price of a Mobile phone on mobile_price_data Dataset

### About Dataset¶
#### # Recently you entered in the mobile manufacturing market. Build a machine learning model that would help you to know the estimated price for your manufactured mobile phones on the basis of various key features and specifications.

#### - Here we are going to Load and Analyse the dataset
#### - Clean the Dataset from Null values and Irrelevant Columns if any existed
#### - Neglecting Unnecessay column out of the Dataset
#### - Remove non-numeric characters(₹) and convert to int dtype in required columns using REGEX.
#### - Feature Encoding, Scaling is done on required column.
#### - Split the Data into Train test split
#### - As this data is a Regression Type - Linear regression Algorithm will be used
#### - Check the accuracy for the Model

### Importing Basic Libraries


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

### Load the data


In [2]:
data = pd.read_csv("mobile_price_data.csv")
data.head()

mobile_name mobile_price mobile_color  dual_sim  \
0    Realme Narzo 10A (So White, 32 GB)       ₹8,999     So White  Dual Sim   
1     Realme Narzo 10A (So Blue, 32 GB)       ₹8,999      So Blue  Dual Sim   
2     Realme Narzo 10A (So Blue, 64 GB)       ₹9,999      So Blue  Dual Sim   
3    Realme Narzo 10A (So White, 64 GB)       ₹9,999     So White  Dual Sim   
4  Realme Narzo 10 (That Green, 128 GB)      ₹11,999   That Green  Dual Sim   

              disp_size         resolution          os  num_cores mp_speed  \
0  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
1  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
2  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
3  16.56 cm (6.52 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   
4   16.51 cm (6.5 inch)  1600 x 720 Pixels  Android 10  Octa Core    2 GHz   

  int_memory   ram                   p_cam              f_cam  \
0      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
1      32 GB  3 GB        12MP + 2MP + 2MP   5MP Front Camera   
2      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
3      64 GB  4 GB        12MP + 2MP + 2MP   5MP Front Camera   
4     128 GB  4 GB  48MP + 8MP + 2MP + 2MP  16MP Front Camera   

                network bluetooth battery_power mob_width mob_height  \
0  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
1  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
2  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
3  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh     75 mm   164.4 mm   
4  4G VOLTE, 4G, 3G, 2G       Yes      5000 mAh   75.4 mm   164.4 mm   

  mob_depth mob_weight  
0   8.95 mm      195 g  
1   8.95 mm      195 g  
2   8.95 mm      195 g  
3   8.95 mm      195 g  
4      9 mm      199 g

In [3]:
#Find the shape of the Data
print("SHAPE: ",data.shape)

SHAPE:  (459, 20)


### Checking for any missing values


In [4]:
print((data.isnull().sum()) / len(data) * 100)

mobile_name      0.0
mobile_price     0.0
mobile_color     0.0
dual_sim         0.0
disp_size        0.0
resolution       0.0
os               0.0
num_cores        0.0
mp_speed         0.0
int_memory       0.0
ram              0.0
p_cam            0.0
f_cam            0.0
network          0.0
bluetooth        0.0
battery_power    0.0
mob_width        0.0
mob_height       0.0
mob_depth        0.0
mob_weight       0.0
dtype: float64


##### As there is no missing values in the Dataset, We are proceeding further...

### Detail Analysis of the dataset


In [5]:
data.describe()

mobile_name mobile_price mobile_color  dual_sim  \
count                               459          459          459       459   
unique                              363          139          165         2   
top     Vivo Z1Pro (Sonic Black, 64 GB)      ₹11,999        Black  Dual Sim   
freq                                  4           19           28       458   

                  disp_size          resolution             os  num_cores  \
count                   459                 459            459        459   
unique                   36                  48             18          3   
top     16.51 cm (6.5 inch)  2340 x 1080 pixels  Android Pie 9  Octa Core   
freq                     51                  66            126        444   

       mp_speed int_memory   ram       p_cam              f_cam  \
count       459        459   459         459                459   
unique       18          7     7          49                 15   
top       2 GHz      64 GB  4 GB  13MP + 2MP  16MP Front Camera   
freq        149        185   160          72                112   

                     network bluetooth battery_power mob_width mob_height  \
count                    459       459           459       459        459   
unique                    18         1            34        70         81   
top     3G, 4G VOLTE, 4G, 2G       Yes      4000 mAh   75.6 mm   156.7 mm   
freq                     139       459           135        35         23   

       mob_depth mob_weight  
count        459        459  
unique        55         62  
top       8.8 mm      191 g  
freq          48         27

In [6]:
data.dtypes

mobile_name      object
mobile_price     object
mobile_color     object
dual_sim         object
disp_size        object
resolution       object
os               object
num_cores        object
mp_speed         object
int_memory       object
ram              object
p_cam            object
f_cam            object
network          object
bluetooth        object
battery_power    object
mob_width        object
mob_height       object
mob_depth        object
mob_weight       object
dtype: object

### Taking the necessary columns in data


In [7]:
necessary = ['mobile_price', 'os','num_cores','mp_speed', 'int_memory', 'ram',
       'p_cam', 'f_cam', 'battery_power','mob_width', 'mob_height',
        'mob_depth','mob_weight']
data = data[necessary]
data.head()

mobile_price          os  num_cores mp_speed int_memory   ram  \
0       ₹8,999  Android 10  Octa Core    2 GHz      32 GB  3 GB   
1       ₹8,999  Android 10  Octa Core    2 GHz      32 GB  3 GB   
2       ₹9,999  Android 10  Octa Core    2 GHz      64 GB  4 GB   
3       ₹9,999  Android 10  Octa Core    2 GHz      64 GB  4 GB   
4      ₹11,999  Android 10  Octa Core    2 GHz     128 GB  4 GB   

                    p_cam              f_cam battery_power mob_width  \
0        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
1        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
2        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
3        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
4  48MP + 8MP + 2MP + 2MP  16MP Front Camera      5000 mAh   75.4 mm   

  mob_height mob_depth mob_weight  
0   164.4 mm   8.95 mm      195 g  
1   164.4 mm   8.95 mm      195 g  
2   164.4 mm   8.95 mm      195 g  
3   164.4 mm   8.95 mm      195 g  
4   164.4 mm      9 mm      199 g

### Remove non-numeric characters(₹) and convert to int dtype of mobile_price column.


In [8]:
data['mobile_price'] = data['mobile_price'].str.replace('[^0-9]', '', regex=True).astype(int)
data.head()

mobile_price          os  num_cores mp_speed int_memory   ram  \
0          8999  Android 10  Octa Core    2 GHz      32 GB  3 GB   
1          8999  Android 10  Octa Core    2 GHz      32 GB  3 GB   
2          9999  Android 10  Octa Core    2 GHz      64 GB  4 GB   
3          9999  Android 10  Octa Core    2 GHz      64 GB  4 GB   
4         11999  Android 10  Octa Core    2 GHz     128 GB  4 GB   

                    p_cam              f_cam battery_power mob_width  \
0        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
1        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
2        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
3        12MP + 2MP + 2MP   5MP Front Camera      5000 mAh     75 mm   
4  48MP + 8MP + 2MP + 2MP  16MP Front Camera      5000 mAh   75.4 mm   

  mob_height mob_depth mob_weight  
0   164.4 mm   8.95 mm      195 g  
1   164.4 mm   8.95 mm      195 g  
2   164.4 mm   8.95 mm      195 g  
3   164.4 mm   8.95 mm      195 g  
4   164.4 mm      9 mm      199 g

In [9]:
data['mp_speed'] = data['mp_speed'].str.replace('[^0-9]', '', regex=True).astype(int)
data['int_memory'] = data['int_memory'].str.replace('[^0-9]', '', regex=True).astype(int)
data['ram'] = data['ram'].str.replace('[^0-9]', '', regex=True).astype(int)
data['battery_power'] = data['battery_power'].str.replace('[^0-9]', '', regex=True).astype(int)
data['mob_width'] = data['mob_width'].str.replace('[^0-9]', '', regex=True).astype(int)
data['mob_height'] = data['mob_height'].str.replace('[^0-9]', '', regex=True).astype(int)
data['mob_depth'] = data['mob_depth'].str.replace('[^0-9]', '', regex=True).astype(int)
data['mob_weight'] = data['mob_weight'].str.replace('[^0-9]', '', regex=True).astype(int)
data.head()

mobile_price          os  num_cores  mp_speed  int_memory  ram  \
0          8999  Android 10  Octa Core         2          32    3   
1          8999  Android 10  Octa Core         2          32    3   
2          9999  Android 10  Octa Core         2          64    4   
3          9999  Android 10  Octa Core         2          64    4   
4         11999  Android 10  Octa Core         2         128    4   

                    p_cam              f_cam  battery_power  mob_width  \
0        12MP + 2MP + 2MP   5MP Front Camera           5000         75   
1        12MP + 2MP + 2MP   5MP Front Camera           5000         75   
2        12MP + 2MP + 2MP   5MP Front Camera           5000         75   
3        12MP + 2MP + 2MP   5MP Front Camera           5000         75   
4  48MP + 8MP + 2MP + 2MP  16MP Front Camera           5000        754   

   mob_height  mob_depth  mob_weight  
0        1644        895         195  
1        1644        895         195  
2        1644        895         195  
3        1644        895         195  
4        1644          9         199

### Feature Encoding on the dataset


In [10]:
df = pd.get_dummies(data, columns=[ 'os','num_cores','p_cam', 'f_cam'])
df.head()

mobile_price  mp_speed  int_memory  ram  battery_power  mob_width  \
0          8999         2          32    3           5000         75   
1          8999         2          32    3           5000         75   
2          9999         2          64    4           5000         75   
3          9999         2          64    4           5000         75   
4         11999         2         128    4           5000        754   

   mob_height  mob_depth  mob_weight  os_Android 10  ...  \
0        1644        895         195              1  ...   
1        1644        895         195              1  ...   
2        1644        895         195              1  ...   
3        1644        895         195              1  ...   
4        1644          9         199              1  ...   

   f_cam_20MP Front Camera  f_cam_24MP Front Camera  f_cam_25MP Front Camera  \
0                        0                        0                        0   
1                        0                        0                        0   
2                        0                        0                        0   
3                        0                        0                        0   
4                        0                        0                        0   

   f_cam_2MP Front Camera  f_cam_32MP + 8MP Dual Front Camera  \
0                       0                                   0   
1                       0                                   0   
2                       0                                   0   
3                       0                                   0   
4                       0                                   0   

   f_cam_32MP Front Camera  f_cam_44MP + 2MP Dual Front Camera  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   f_cam_5MP Front Camera  f_cam_8MP Dual Front Camera  f_cam_8MP Front Camera  
0                       1                            0                       0  
1                       1                            0                       0  
2                       1                            0                       0  
3                       1                            0                       0  
4                       0                            0                       0  

[5 rows x 94 columns]

### Feature scaling on dataset


In [11]:
from sklearn.preprocessing import RobustScaler
rs = RobustScaler()
rs

RobustScaler()

In [12]:
col=['mobile_price','mp_speed', 'int_memory', 'battery_power','mob_width', 'mob_height',
        'mob_depth','mob_weight']
df[col] = rs.fit_transform(df[col])
df.head()

mobile_price  mp_speed  int_memory  ram  battery_power  mob_width  \
0     -0.687774 -0.761905   -0.333333    3           1.94  -0.100724   
1     -0.687774 -0.761905   -0.333333    3           1.94  -0.100724   
2     -0.562633 -0.761905    0.000000    4           1.94  -0.100724   
3     -0.562633 -0.761905    0.000000    4           1.94  -0.100724   
4     -0.312351 -0.761905    0.666667    4           1.94  -0.000443   

   mob_height  mob_depth  mob_weight  os_Android 10  ...  \
0    0.001628   1.117729    0.295082              1  ...   
1    0.001628   1.117729    0.295082              1  ...   
2    0.001628   1.117729    0.295082              1  ...   
3    0.001628   1.117729    0.295082              1  ...   
4    0.001628  -0.109418    0.426230              1  ...   

   f_cam_20MP Front Camera  f_cam_24MP Front Camera  f_cam_25MP Front Camera  \
0                        0                        0                        0   
1                        0                        0                        0   
2                        0                        0                        0   
3                        0                        0                        0   
4                        0                        0                        0   

   f_cam_2MP Front Camera  f_cam_32MP + 8MP Dual Front Camera  \
0                       0                                   0   
1                       0                                   0   
2                       0                                   0   
3                       0                                   0   
4                       0                                   0   

   f_cam_32MP Front Camera  f_cam_44MP + 2MP Dual Front Camera  \
0                        0                                   0   
1                        0                                   0   
2                        0                                   0   
3                        0                                   0   
4                        0                                   0   

   f_cam_5MP Front Camera  f_cam_8MP Dual Front Camera  f_cam_8MP Front Camera  
0                       1                            0                       0  
1                       1                            0                       0  
2                       1                            0                       0  
3                       1                            0                       0  
4                       0                            0                       0  

[5 rows x 94 columns]

### Split the data into training and testing sets


In [13]:
from sklearn.model_selection import train_test_split
X = df.drop('mobile_price', axis=1)
y = df['mobile_price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Build and train a linear regression model


In [14]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr

LinearRegression()

In [15]:
lr.fit(X_train, y_train)

LinearRegression()

In [16]:
y_pred = lr.predict(X_test)


### Calculate the Accuracy


In [17]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print("r2_score:", r2)

r2_score: 0.9403416240000377


## The Machine Learning Model Predicts the dataset with 94% of accuracy